### PR TITLE
i8: add DotProductBatch register-blocked quantized matrix-vector kernel

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

`DotProductBatch(results []int32, rows [][]int8, vec []int8)` scores many int8 weight rows against one shared int8 activation vector into int32 (`results[i] = DotProduct(rows[i], vec)`). It is the quantized GEMM matrix-vector core and the highest-impact remaining Tier 1 item of the int8 surface.

A fused 4-row kernel keeps `vec` resident across the group instead of re-streaming it per row: on AVX2 it sign-extends with `VPMOVSXBW` and multiply-accumulates with `VPMADDWD`/`VPADDD`; on ARM64 it uses `SDOT` into four accumulators that share the `vec` register loaded once per 16-byte block. Ragged groups (any row shorter than `vec`) and the trailing rows past the last full group fall back to the shared per-row dispatch, which itself vectorizes each row through the existing `DotProduct` kernel. int32 addition wraps associatively, so any row grouping is bit-identical to the scalar Go reference, which makes parity the same as bit-exactness (there is no float rounding contract to preserve).

The dispatch gates the fused kernel on `len(rows) >= 4` and `vecLen >= 16` (and `FEAT_DotProd` on ARM64); everything below that, plain NEON, and the pure-Go platform take the per-row fallback. An empty `vec` zeroes `results`, matching the empty-row rule. The operation is allocation-free.

Measured roughly 1.6 to 2.6x over a per-row `DotProduct` loop across a dims x rows grid, with zero allocations, on an i7-1260P (AVX2) and a Raspberry Pi 5 Cortex-A76 (NEON SDOT); the win is largest at the small-to-medium activation widths typical of quantized inference and converges toward the memory-bound per-row loop at very large dims.

## Related Issues

Advances the int8 surface tracking issue #132 (Tier 1, quantized inference). The umbrella issue stays open for the remaining Tier 1/2/3 items.

## Test Plan

- [x] Parity against an independent scalar oracle across vecLen and row-count sweeps that straddle the dispatch gates, the 16-wide loop, the AVX2 8-wide prelude, and the scalar tail
- [x] Ragged, nil, empty, over-long, and single-element rows; full-group kernel path; trailing empty rows; length clamp and trailing-untouched results; empty-vec zeroing
- [x] int32 two's-complement wraparound forced past 2^31 and checked against the wrapping oracle
- [x] `testing.AllocsPerRun == 0`
- [x] Differential fuzzing against the oracle (millions of executions, seeded to reach the SIMD kernel path)
- [x] `asmcheck` cross-checks the new `SDOT`/`ADDV` WORD encodings against `arm64asm`; reserved-register and ISA/FMA gates pass
- [x] Built and tested on amd64 (AVX2) and on ARM64 hardware (Cortex-A76, FEAT_DotProd)
- [x] `go vet`, `golangci-lint run ./...`, `gofmt` all clean

## Summary by CodeRabbit

* **New Features**
  * Added `DotProductBatch` for computing multiple int8 dot products against a shared vector.
  * Supports varied row lengths, empty inputs, limited result ranges, and allocation-free execution.
  * Added documentation and an example demonstrating batched dot-product results.

* **Performance**
  * Added optimized batched processing for supported AVX2 and ARM64 DotProd hardware, with portable fallbacks.

* **Tests**
  * Added comprehensive correctness, fuzz, edge-case, and benchmark coverage.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -770,6 +771,12 @@ i8.AbsDiff(dst, a, b)     // saturating |a - b|, clamped to [0, 127]
 
 dot := i8.DotProduct(a, b) // int32-accumulated sum(a[i]*b[i])
 sum := i8.Sum(a)           // int32-accumulated sum
+
+// Batched matrix-vector: one int32 dot product per weight row against a shared
+// activation vector, with the vector kept resident across the rows.
+rows := [][]int8{ /* int8 weight rows, each len(vec) long */ }
+results := make([]int32, len(rows))
+i8.DotProductBatch(results, rows, a) // results[i] = DotProduct(rows[i], a)
```
[the remark this answers is not in the document]
Comment (tphakala) [reply README.md:779]: Fixed in cbce954. The example comment now says `len(a)` to match the activation vector actually in scope in that snippet.

File: i8/i8.go [hunk 2]
```diff
@@ -194,6 +197,26 @@ func DotProduct(a, b []int8) int32 {
...
+// n = min(len(results), len(rows)). Each row contributes min(len(rows[i]),
+// len(vec)) elements (the same clamp DotProduct applies); an empty row, or an
+// empty vec, scores 0. It is the matrix-vector core of
+// quantized inference (int8 weight rows against an int8 activation vector).
+// Batching the rows keeps vec resident across them (in registers under the
+// 4-row kernel, in L1 otherwise) instead of re-streaming it per DotProduct call.
+//
+// results, rows and vec must not overlap, and rows and vec are read-only.
+// Trailing capacity in results beyond n is left untouched. The call allocates
+// nothing.
```
[the remark this answers is not in the document]
Comment (tphakala) [reply i8/i8.go]: Fixed in cbce954. Reworded the contract: rows and vec are read-only and may overlap (a row can even be vec itself, as the wraparound test relies on), so the old no-overlap requirement was wrong. results is a distinct []int32 slice. Also changed "capacity" to "elements from index n onward" since those are slice elements, not spare capacity.

**Comment by tphakala** [comment]
Thanks, applied the nitpick in cbce954: the allocation test now uses length 63 so the fused kernels' scalar-tail residue is exercised allocation-free too, not just the 16-aligned path.
